Generic extension support

Selmar opened the pull request:
Two issues remaining:
- The lights data exists twice on export; once outside of the gltf specification and once in the extensions.
- A lot of objects are still missing the extensions property.

syoyo commented:
Thank you for PR. Will review soon.

> The lights data exists twice on export; once outside of the gltf specification and once in the extensions.

Do you have minimal reproducing test data for this?

Selmar commented:
It can't actually load the lights extension correctly; it's good you asked me to. The nesting concerns I had earlier were valid concerns.

I think we should use a `map<string, Value>` instead of a `map<string, ParameterMap>`.

Selmar commented:
Here is my test case, using the cube example. `modeltest.gltf` is the imported file, `lightsTest.gltf` is the re-serialized file.

[lightsTest.zip](https://github.com/syoyo/tinygltf/files/1818936/lightsTest.zip)

syoyo commented:
I have add some modifications and now lights are serialized under `KHR_lights_cmn` as in `modeltest.gltf` in this branch : https://github.com/syoyo/tinygltf/tree/Selmar-generic_extension_support

````
  "extensions": {
    "KHR_lights_cmn": {
      "lights": [
        {
          "color": [
            0,
            1,
            2,
            3
          ],
          "name": "testLight",
          "type": "fake"
        }
      ]
    }
```

Selmar commented:
Alright. I suppose you want to change it to using a Value instead of ParameterMap before you merge it? And perhaps some input from people who use the existing extensions objects.

syoyo commented:
> I suppose you want to change it to using a Value instead of ParameterMap before you merge it? 

Yes! you are right.

Selmar commented on tiny_gltf.h:
```diff
@@ -1704,7 +1705,7 @@ static bool ParseExtensionsProperty(ExtensionMap *ret, std::string* err, const j
     for (; itVal != itValEnd; itVal++) {
       Parameter param;
-      if (ParseParameterProperty(&param, err, values_object, itVal.key(), false)) {
+      if (ParseParameterProperty(&param, err, itVal.value(), itVal.key(), false)) {
```
I have just found that this change actually broke the extension parsing.

I will be changing the ExtensionMap to use Value objects instead, so this'll be fixed along the way.

Selmar commented:
- changed: parsing integer types to `Value` with type `INT_TYPE`
  - motivation: we should either merge `int_value` and `number_value` or treat them differently
    - if we keep a separation, I suggest to rename `NUMBER_TYPE` to `DECIMAL_TYPE` for clarity, because  integers are numbers too
  - previously, if a json parsed value was an integer, it was parsed as a double regardless
- parsing all json types except for `discarded`
  - e.g. for extras
  - `json::value_t::discarded` is ignored
  - conversion to `Value` with type `BINARY_TYPE` not done (not sure what the intention was)
- `Value` entries for `object` and `array` types are not added if they are `empty` or `null`
- `ParseExtrasProperty` returns false if the parsed property is `NULL_TYPE`
  - previously always returned true
  - doesn't seem to be used anywhere
- Serialization also tested and seems to be functional
  - there is currently still a difference in code design between parsing and serialization of value objects (see `SerializeValue`, `ParseExtrasProperty` and `ParseJsonAsValue`)
  - serializing `BINARY_TYPE` is not done
- Added parsing of extension and extras to Texture
  - we will use it for .dds files (`MSFT_texture_dds`)

syoyo commented:
Thanks! I have merged your commits into https://github.com/syoyo/tinygltf/tree/Selmar-generic_extension_support

Will soon `update loader_example.cc` then merge it into `devel`

Selmar commented:
merged, but not tested

syoyo commented:
I have fixed `loader_example` and added initial support for printing extensions. It looks working so merged into `devel`.
